torch/cuda/memory_viz.py

```diff
@@ -0,0 +1,188 @@
+            frames = h['frames']
+            if frames:
+                frame_s = ';'.join([_frame_fmt(f) for f in reversed(frames)])
+            else:
+                frame_s = "<non-python>"
+            f.write(f'{prefix};{b["state"]};{frame_s} {sz}\n')
+        gaps = b['size'] - accounted_for_size
+        if gaps:
+            f.write(f'{prefix};{b["state"]};<gaps> {gaps}\n')
+
+def segments(snapshot, format_flamegraph=format_flamegraph):
+    f = io.StringIO()
+    for seg in snapshot:
+        prefix = f'stream_{seg["stream"]};seg_{seg["address"]}'
+        _write_blocks(f, prefix, seg['blocks'])
+    return format_flamegraph(f.getvalue())
+
+def memory(snapshot, format_flamegraph=format_flamegraph):
+    f = io.StringIO()
+    for seg in snapshot:
+        prefix = f'stream_{seg["stream"]}'
+        _write_blocks(f, prefix, seg['blocks'])
+    return format_flamegraph(f.getvalue())
+
+def compare(before, after, format_flamegraph=format_flamegraph):
+    def _seg_key(seg):
+        return (seg['address'], seg['total_size'])
+
+    def _seg_info(seg):
+        return f'stream_{seg["stream"]};seg_{seg["address"]}'
+
+    f = io.StringIO()
+
+    before_segs = set(_seg_key(seg) for seg in before)
+    after_segs = set(_seg_key(seg) for seg in after)
+
+    print(f'only_before = {list(a for a,_ in (before_segs - after_segs))}')
+    print(f'only_after = {list(a for a,_ in (after_segs - before_segs))}')
+
+    for seg in before:
+        if _seg_key(seg) not in after_segs:
+            _write_blocks(f, f'only_before;{_seg_info(seg)}', seg['blocks'])
+
+    for seg in after:
+        if _seg_key(seg) not in before_segs:
+            _write_blocks(f, f'only_after;{_seg_info(seg)}', seg['blocks'])
+
+    return format_flamegraph(f.getvalue())
+
+class Bytes:
+    def __init__(self, value):
+        self.value = value
+
+    def __add__(self, rhs):
+        return Bytes(self.value + rhs)
+
+    def __repr__(self):
+        num = self.value
+        # https://stackoverflow.com/questions/1094841/get-human-readable-version-of-file-size
+        for unit in ["", "Ki", "Mi", "Gi", "Ti", "Pi", "Ei", "Zi"]:
+            if abs(num) < 1024.0:
+                return f"{num:3.1f}{unit}B"
+            num /= 1024.0
+        return f"{num:.1f}YiB"
+
+def stats(snapshot):
+    result : Dict[str, Any] = {}
+    result['segments'] = len(snapshot)
+    result['total_size'] = Bytes(0)
+    for seg in snapshot:
+        total_size = 0
+        for b in seg['blocks']:
+            if b['state'] not in result:
+                result[b['state']] = Bytes(0)
+            total_size += b['size']
+            result[b['state']] += b['size']
+        assert seg['total_size'] == total_size
+        result['total_size'] += total_size
+    return result
+
+
+if __name__ == "__main__":
+    import os.path
+    thedir = os.path.realpath(os.path.dirname(__file__))
+    if thedir in sys.path:
+        # otherwise we find cuda/random.py as random...
```

Comment:
Where is that a problem?

Reply:
when you run this script right out of the repo. I use it on my Mac without even building PyTorch. We'll organize this better after figuring out what the final toolset will look like.

Reply:
Ok